From 3e8aadf321c9a5eb2b4982a24d62283b89ac5b22 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 27 Mar 2025 18:55:04 +0100
Subject: [PATCH 1645/1695] plat-stm32mp2: clean WKUP IRQ management

Clean the PWR WKUP management, as it is described in reference manual:
- clear flag before enable to avoid false interrupt
- don't disable PWR interruption for mask operations because
  mask the parent interrupt is enough for interrupted framework
  and this patch avoids to lost a PWR interruption between
  call of mask and unmask function, when interruption are disabled.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git fef025056768156d19ed88711cba4857b72c166b plat-stm32mp2: clean WKUP IRQ management]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I858dfb5bb026c937af82b08db84c778cc84d2ae6
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/447260
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Reviewed-by: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
index a95e8c91c..f56443c1b 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_irq.c
@@ -141,6 +141,8 @@ static void stm32mp25_pwr_itr_enable_nolock(size_t it)
 {
 	VERBOSE_PWR("Pwr irq enable %zu", it);
 
+	/* Clear flag before enable to avoid false interrupt */
+	io_setbits32(pwr_wkupcr_base(it), WKUPC);
 	io_setbits32(pwr_wkupcr_base(it), WKUPENCPU1);
 }
 
@@ -228,7 +230,6 @@ static void stm32mp25_pwr_op_mask(struct itr_chip *chip __unused, size_t it)
 
 	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
 	priv->itr_mask_bitmask |= BIT(it);
-	stm32mp25_pwr_itr_disable_nolock(it);
 	interrupt_mask(priv->parent_hdl->chip, priv->parent_hdl->it);
 	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
 }
@@ -241,7 +242,6 @@ static void stm32mp25_pwr_op_unmask(struct itr_chip *chip __unused, size_t it)
 
 	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
 	priv->itr_mask_bitmask &= ~BIT(it);
-	stm32mp25_pwr_itr_enable_nolock(it);
 	if (!priv->itr_mask_bitmask)
 		interrupt_unmask(priv->parent_hdl->chip, priv->parent_hdl->it);
 	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
-- 
2.47.2

